builder/amazon/common/step_create_ssm_tunnel.go

```diff
@@ -0,0 +1,135 @@
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	dst, src := strconv.Itoa(s.DstPort), strconv.Itoa(l.Port)
+	params := map[string][]*string{
+		"portNumber":      []*string{aws.String(dst)},
+		"localPortNumber": []*string{aws.String(src)},
+	}
+	l.Close()
+
+	instance, ok := state.Get("instance").(*ec2.Instance)
+	if !ok {
+		err := fmt.Errorf("error encountered in obtaining target instance id for SSM tunnel")
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	s.instanceId = aws.StringValue(instance.InstanceId)
+	ssmconn := ssm.New(s.AWSSession)
+	input := ssm.StartSessionInput{
+		DocumentName: aws.String("AWS-StartPortForwardingSession"),
+		Parameters:   params,
+		Target:       aws.String(s.instanceId),
+	}
+
+	ui.Message(fmt.Sprintf("Starting PortForwarding session to instance %q on local port %q to remote port %q", s.instanceId, src, dst))
+	var output *ssm.StartSessionOutput
+	err = retry.Config{
+		ShouldRetry: func(err error) bool { return isAWSErr(err, "TargetNotConnected", "") },
+		RetryDelay:  (&retry.Backoff{InitialBackoff: 200 * time.Millisecond, MaxBackoff: 60 * time.Second, Multiplier: 2}).Linear,
+	}.Run(ctx, func(ctx context.Context) error {
+		output, err = ssmconn.StartSessionWithContext(ctx, &input)
+		return err
+	})
+
+	if err != nil {
+		err = fmt.Errorf("error encountered in starting session for instance %q: %s", s.instanceId, err)
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+	s.ssmSession = output
+
+	// AWS session-manager-plugin requires a valid session be passed in JSON
+	sessionDetails, err := json.Marshal(s.ssmSession)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session details", err)
+		return multistep.ActionHalt
+	}
+
+	sessionParameters, err := json.Marshal(input)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session parameter details", err)
+		return multistep.ActionHalt
+	}
+
+	driver := SSMDriver{Ui: ui}
+	// sessionDetails, region, "StartSession", profile, paramJson, endpoint
+	region := aws.StringValue(s.AWSSession.Config.Region)
+	// how to best get Profile name
+	if err := driver.StartSession(string(sessionDetails), region, "", string(sessionParameters), ssmconn.Endpoint); err != nil {
+		err = fmt.Errorf("error encountered in establishing a tunnel with the session-manager-plugin: %s", err)
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	ui.Message(fmt.Sprintf("PortForwarding session to instance %q established!", s.instanceId))
+	state.Put("sessionPort", l.Port)
+
+	return multistep.ActionContinue
+}
+
+func (s *StepCreateSSMTunnel) Cleanup(state multistep.StateBag) {
+	if s.ssmSession == nil {
+		return
+	}
+
+	ui := state.Get("ui").(packer.Ui)
+	ssmconn := ssm.New(s.AWSSession)
+	_, err := ssmconn.TerminateSession(&ssm.TerminateSessionInput{SessionId: s.ssmSession.SessionId})
```

Comment:
> Can this be retried as well? If there's a situation where the session takes some time to "unblock" before terminating, it can be annoying to terminate manually every time.  If this is a real possible situation.

That's a good question. It is possible that it is retryable so let me check and update the code if it is. In reading the docs the session will automatically terminate itself if it is idle for more than 20 minutes.

Reply:
> Another question does aws allow to deregister the ami instance if there's a port forward session going on? What happens if someone kills the build in the middle and its not able to terminate the session, would packer be able to clean the instance?

This is a great question and one to test. If the instance is destroyed before calling terminate I think the session will be automatically killed because the tunnel is between the local machine and the ssm-agent running on the remote instance. If the ssm-agent is no longer active I believe that the Systems Manager service will automatically terminate the session and stop the port forwarding. 

If in the case the instance is still active and the Packer build has been cancelled (destroy step was also cancelled) then the session with auto terminate if there is no activity. I think :smile:

Reply:
Nice! So I guess it wouldn't be a problem for both situations, right?  Thankss

Reply:
> Can this be retried as well? 

Doesn't seem like this is the case. TerminateSession returns two non-retryable errors, at least from what is documented, DoesNotExistException and InternalServerError.